### PR TITLE
[GEOT-7486] CSS literal single space squashed to empty string

### PR DESCRIPTION
**Pull request by petersmythe**
[GEOT-7486](https://osgeo-org.atlassian.net/browse/GEOT-7486) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

Fix now produces

`<ogc:Literal><![CDATA[ ]]></ogc:Literal>`

instead of the problematic `<ogc:Literal> </ogc:Literal>`

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [x] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [x] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [x] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [x] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

**Comment by aaime** [comment]
The change looks reasonable, but without a test it cannot be considered for merge.

**Comment by petersmythe** [comment]
Thanks @aaime , I have had a look at `TransformerBaseTest` and I cannot see how I would implement such a test.  Could you possibly give me some guidance, please?

**Comment by aaime** [comment]
You should be looking at SLDTransformerTest, and do something similar to [testLabelCDataStart](https://github.com/geotools/geotools/blob/64f714ec7c36aadaf62310fd941d89388c4ab57b/modules/library/xml/src/test/java/org/geotools/xml/styling/SLDTransformerTest.java#L1536), but making sure the code path just modified is hit instead. 
I'd take that test and change it so that you build a style using the WKT Parsing function, that should keep you close to your initial finding and make sure that use case cannot regress in the future.

**Comment by petersmythe** [comment]
Thanks for the hint, Andrea, I believe this tests the change properly

File: modules/library/xml/src/main/java/org/geotools/xml/transform/TransformerBase.java [hunk 1]
```diff
@@ -876,7 +876,11 @@ protected void element(String element, String content, Attributes atts) {
             start(element, atts);
 
             if (content != null) {
-                chars(content);
+                if (content.matches("^\\s+$")) {
```
Comment (aaime) [inline modules/library/xml/src/main/java/org/geotools/xml/transform/TransformerBase.java:879]: So all good, but I'm worried performance-wise about this bit... every time it's invoked, it recompiles the Java Pattern and then executes it, which is going to slow down a lot everything that is based on TransformerBase, like large OGC Capabilities documents.

Wondering if we can compromise a bit, and go with a less generic, but more efficient version:

```suggestion
                if (content.indexOf(" ") != -1) {
```

Comment (aaime) [reply modules/library/xml/src/main/java/org/geotools/xml/transform/TransformerBase.java:879]: Actually wait, it can be done even more efficiently.. because XML squashes the trailing and leading spaces, but not the ones in between (no need for CDATA in that case), so one should use something like this instead, and without sacrificing generality:

```
int length = content.length();
if (length > 0 && (Character.isWhiteSpace(content.charAt(0)) || Character.isWhiteSpace(content.charAt(length - -1))) { ...
```

Comment (petersmythe) [reply modules/library/xml/src/main/java/org/geotools/xml/transform/TransformerBase.java:879]: That is very true.  Thank you for the improvement.

I took the cue from https://github.com/geotools/geotools/blob/153cc712c0d7178989effeafca03238fb27f2f18/modules/library/xml/src/main/java/org/geotools/xml/styling/SLDTransformer.java#L343 but that code is only executed once when updating the (CSS) style.

Comment (petersmythe) [reply modules/library/xml/src/main/java/org/geotools/xml/transform/TransformerBase.java:879]: > Actually wait, it can be done even more efficiently.. because XML squashes the trailing and leading spaces, but not the ones in between (no need for CDATA in that case), so one should use something like this instead, and without sacrificing generality:
> 
> ```
> int length = content.length();
> if (length > 0 && (Character.isWhiteSpace(content.charAt(0)) || Character.isWhiteSpace(content.charAt(length - -1)) { ...
> ```

`length - -1` or `length -1`?

**Comment by aaime** [comment]
The backport to `29.x` failed:
```
The process '/usr/bin/git' failed with exit code 1
```
<details>
<summary>stderr</summary>

```
error: could not apply d0d9bf9dc8... Add unit test for whitespace CDATA
hint: After resolving the conflicts, mark them with
hint: "git add/rm <pathspec>", then run
hint: "git cherry-pick --continue".
hint: You can instead skip this commit with "git cherry-pick --skip".
hint: To abort and get back to the state before "git cherry-pick",
hint: run "git cherry-pick --abort".

```
</details>

<details>
<summary>stdout</summary>

```
Auto-merging modules/library/xml/src/main/java/org/geotools/xml/transform/TransformerBase.java
[backport-4537-to-29.x c3d849b5bb] [GEOT-7486] CSS literal single space squashed to empty string
 Author: Peter Smythe <peter@afrigis.co.za>
 Date: Fri Nov 3 13:44:12 2023 +0200
 1 file changed, 5 insertions(+), 1 deletion(-)
Auto-merging modules/library/xml/src/main/java/org/geotools/xml/transform/TransformerBase.java
[backport-4537-to-29.x 653de652ea] [GEOT-7486] CSS literal single space squashed to empty string
 Author: Peter Smythe <peter@afrigis.co.za>
 Date: Fri Nov 3 14:45:27 2023 +0200
 1 file changed, 1 insertion(+), 1 deletion(-)
Auto-merging modules/library/xml/src/test/java/org/geotools/xml/styling/SLDTransformerTest.java
CONFLICT (content): Merge conflict in modules/library/xml/src/test/java/org/geotools/xml/styling/SLDTransformerTest.java

```
</details>

To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Create a new working tree
git worktree add .worktrees/backport-29.x 29.x
# Navigate to the new working tree
cd .worktrees/backport-29.x
# Create a new branch
git switch --create backport-4537-to-29.x
# Cherry-pick the merged commit of this pull request and resolve the conflicts
git cherry-pick a44a78d469762e99862000e8d73b8e0175fec076,65a5203d84e7d117e588dd950f47fde6e3b831cd,d0d9bf9dc8eb359a5afe04ea97e36f69a6ca8cd5,96472dd15da2ada2639fcae867ea882f089f777e
# Push it to GitHub
git push --set-upstream origin backport-4537-to-29.x
# Go back to the original working tree
cd ../..
# Delete the working tree
git worktree remove .worktrees/backport-29.x
```
Then, create a pull request where the `base` branch is `29.x` and the `compare`/`head` branch is `backport-4537-to-29.x`.
